s3-sink-connector/src/main/java/io/aiven/kafka/connect/s3/S3SinkTask.java

```diff
@@ -112,39 +134,118 @@ public void put(final Collection<SinkRecord> records) {
         Objects.requireNonNull(records, "records cannot be null");
         LOGGER.info("Processing {} records", records.size());
         records.forEach(recordGrouper::put);
+        if (!isKeyRecordGrouper) {
+            recordGrouper.records().forEach(this::writeToS3);
+        }
     }
 
+    /**
+     * Flush is used alongside the KeyRecordGroupers to initiate and complete file writes to S3. When not using a key
+     * record grouper, the S3 upload will be initiated by the put command and flush will be used to write the files and
+     * roll over the files if any records remain in the record grouper for completion.
+     *
+     * @param offsets
+     *            the latest offset sent to put and that is now ready to be flushed.
+     */
     @Override
     public void flush(final Map<TopicPartition, OffsetAndMetadata> offsets) {
         try {
-            recordGrouper.records().forEach(this::flushFile);
+            recordGrouper.records().forEach(this::flushToS3);
```

Comment:
Talking to @AnatolyPopov we want to do two things 
1) Add documentation in README for S3 bucket lifecycle to remove incomplete multipart upload.
2) Test if upload is aborted due to malformed or IOException that all the records are retried when connector restarts/retries.